### ISSUE TITLE
Allow running code after spew() iterates, but before it signals we're done

### ISSUE THREAD

**Issue by vitorbaptista** [frictionlessdata/datapackage-pipelines#98]
Consider a pipeline that ends with 2 processors:

* ...
* dump.to_zip
* upload_zip

`dump.to_zip` writes a datapackage as a zip in the local filesystem, which `upload_zip` loads and uploads to some service. This means that they have a dependency: `upload_zip` can't run before `dump.to_zip` is finished.

The design of datapackage-pipelines is to work with streaming data. Every processor runs in parallel, taking the data rows as they come. The current pattern that we use to make sure the previous processors have finished running is to call `spew()` to pass the data on to the next processors (modified or not), and then run our code. This works fine if there's only one processor that uses this pattern, but not if there's more than one. Consider this simplified code for `dump.to_zip` and `upload_zip`.

* `dump.to_zip`
```python
zip_file = zipfile.ZipFile(out_file)

spew(
    datapackage,
    resource_iter  # This iterator will write the files to `zip_file`
)

# After spew() has finished, it'll unblock the following processor's spew() calls
# so the code after their spew() calls will start running.

zip_file.close()
```

* `upload_zip`
```python
# Block until the previous processor's spew() calls have finished
spew(datapackage, resource_iter)

# Load zip from path and upload it
upload_zip(params['in_file'])
```

This doesn't guarantee that `dump.to_zip` will call `zip_file.close()` before `upload_zip` read the zip file, so it might be reading an invalid zip from disk (because `dump.to_zip` is still writing the contents). We have a race condition.

This race condition happens because the processors are communicating outside `datapackage-pipelines`, using the filesystem. As `dpp` doesn't know about this, it has no way to coordinate their communication. To avoid this, we must get the code in `dpp`.

My proposal is to add a `finalizer` parameter to `spew()` where you can send finalization code to be executed after `spew()` has done processing the incoming data, but before it has signaled to the next processors in the pipeline that we're finished. This way, the code for (for example) `dump.to_zip` would become:

```python
zip_file = zipfile.ZipFile(out_file)

spew(
    datapackage,
    resource_iter,
    finalizer=lambda: zip_file.close()
)
```

I have implemented this pattern on https://github.com/openspending/datapackage-pipelines-fiscal/commit/391048159ad21e647e7e5b4b316467848e3d2bf3, although as that's outside `dpp`, I had to do it in a hackish way.

**Comment by vitorbaptista** [comment]
@akariv Sorry, I commited this directly to master by mistake. Could you take a look on https://github.com/frictionlessdata/datapackage-pipelines/commit/d591baa2c4ed340e9557ec6056aa672e05938693 ? We can revert it to go through the PR review process, or do any changes you'd like in a separate PR.
